File: LTE_record_process/+ltebuffer/CsiFrameAssembler.m
classdef CsiFrameAssembler < handle
%CSIFRAMEASSEMBLER Assemble ten consecutive CSI subframes into one frame.

    properties (SetAccess = private)
        Epoch = NaN
        BufferedSubframes = 0
        EmittedFrames = 0
        DroppedPartialFrames = 0
    end

    properties (Access = private)
        Buffer
        FirstMeta
        ExpectedSequence = NaN
    end

    methods
        function result = push(obj, packet)
            obj.validatePacket(packet);
            meta = packet.Meta;
            if ~isequal(obj.Epoch, meta.Epoch)
                obj.reset(meta.Epoch, 'new-epoch');
            end
            if obj.BufferedSubframes > 0 && ...
                    meta.Sequence ~= obj.ExpectedSequence
                obj.discardPartial();
            end
            if obj.BufferedSubframes > 0 && ...
                    meta.SubframeNumber ~= obj.BufferedSubframes
                obj.discardPartial();
            end
            if obj.BufferedSubframes == 0
                if meta.SubframeNumber ~= 0
                    result = obj.emptyResult();
                    return;
                end
                obj.allocate(packet.Data);
                obj.FirstMeta = meta;
            end

            columns = meta.SubframeNumber*2 + (1:2);
            obj.Buffer.G1(:, columns, :, :) = packet.Data.G1;
            obj.Buffer.G2(:, columns, :, :) = packet.Data.G2;
            obj.Buffer.DynamicG1(:, columns, :, :) = packet.Data.DynamicG1;
            obj.Buffer.DynamicG2(:, columns, :, :) = packet.Data.DynamicG2;
            obj.BufferedSubframes = obj.BufferedSubframes + 1;
            obj.ExpectedSequence = meta.Sequence + 1;

            result = obj.emptyResult();
            if meta.SubframeNumber ~= 9
                return;
            end
            frameMeta = obj.FirstMeta;
            frameMeta.RawEndSample0 = meta.RawEndSample0;
            frameMeta.EndSequence = meta.Sequence;
            frameMeta.SubframeNumber = [];
            result.Available = true;
            result.Packet = struct('Data', obj.Buffer, ...
                'Meta', frameMeta, ...
                'Quality', struct('Complete', true, ...
                'SubframeCount', obj.BufferedSubframes));
            obj.EmittedFrames = obj.EmittedFrames + 1;
            obj.clearPartial();
        end

        function reset(obj, epoch, reason) %#ok<INUSD>
            if nargin < 2
                epoch = NaN;
            end
            if obj.BufferedSubframes > 0
                obj.DroppedPartialFrames = obj.DroppedPartialFrames + 1;
            end
            obj.Epoch = epoch;
            obj.Buffer = struct();
            obj.FirstMeta = struct();
            obj.BufferedSubframes = 0;
            obj.ExpectedSequence = NaN;
        end

        function status = getStatus(obj)
            status = struct('State', 'ready', 'Ready', true, ...
                'Epoch', obj.Epoch, ...
                'BufferedSubframes', obj.BufferedSubframes, ...
                'EmittedFrames', obj.EmittedFrames, ...
                'DroppedPartialFrames', obj.DroppedPartialFrames);
        end
    end

    methods (Access = private)
        function allocate(obj, data)
            obj.Buffer = struct( ...
                'G1', complex(zeros(size(data.G1, 1), 20, ...
                size(data.G1, 3), size(data.G1, 4), 'like', data.G1)), ...
                'G2', complex(zeros(size(data.G2, 1), 20, ...
                size(data.G2, 3), size(data.G2, 4), 'like', data.G2)), ...
                'DynamicG1', complex(zeros(size(data.DynamicG1, 1), 20, ...
                size(data.DynamicG1, 3), size(data.DynamicG1, 4), ...
                'like', data.DynamicG1)), ...
                'DynamicG2', complex(zeros(size(data.DynamicG2, 1), 20, ...
                size(data.DynamicG2, 3), size(data.DynamicG2, 4), ...
                'like', data.DynamicG2)));
        end

        function discardPartial(obj)
            if obj.BufferedSubframes > 0
                obj.DroppedPartialFrames = obj.DroppedPartialFrames + 1;
            end
            obj.clearPartial();
        end

        function clearPartial(obj)
            obj.Buffer = struct();
            obj.FirstMeta = struct();
            obj.BufferedSubframes = 0;
            obj.ExpectedSequence = NaN;
        end

        function validatePacket(~, packet)
            required = {'G1', 'G2', 'DynamicG1', 'DynamicG2'};
            if ~isstruct(packet) || ~isfield(packet, 'Data') || ...
                    ~isfield(packet, 'Meta') || ...
                    ~all(isfield(packet.Data, required))
                error('ltebuffer:CsiFrameAssembler:InvalidPacket', ...
                    'Input is not a valid CSI subframe packet.');
            end
            metaFields = {'Epoch', 'Sequence', 'SubframeNumber'};
            if ~all(isfield(packet.Meta, metaFields))
                error('ltebuffer:CsiFrameAssembler:InvalidMetadata', ...
                    'CSI subframe metadata is incomplete.');
            end
        end

        function result = emptyResult(obj) %#ok<MANU>
            result = struct('Available', false, 'Packet', struct());
        end
    end
end

File: LTESen_python/ltesen/ltebuffer/csi_frame_assembler.py
"""Assemble ten consecutive CSI subframes into a complete LTE frame."""

from __future__ import annotations

import math
from typing import Any

import numpy as np

from ..ltepipe import Message, Module, Packet, Result
from ._array_shapes import as_csi_4d, complex_zeros_like


class CsiFrameAssembler(Module):
    """Collect subframes 0..9, rejecting epochs and sequence gaps."""

    _FIELDS = ("g1", "g2", "dynamic_g1", "dynamic_g2")
    _MATLAB_FIELDS = {
        "g1": "G1",
        "g2": "G2",
        "dynamic_g1": "DynamicG1",
        "dynamic_g2": "DynamicG2",
    }

    def __init__(self) -> None:
        super().__init__("frame_assembler", "csi-subframe", "csi-frame")
        self.epoch: int | float | None = None
        self.buffered_subframes = 0
        self.emitted_frames = 0
        self.dropped_partial_frames = 0
        self._buffer: dict[str, np.ndarray] | None = None
        self._first_meta: dict[str, Any] | None = None
        self._expected_sequence: int | float | None = None

    def process(self, message: Message) -> Result:
        self.validate_input(message)
        if not message.has_packet:
            return Result.forward(message)
        assert message.packet is not None
        self._validate_packet(message.packet)
        packet = message.packet
        meta = _normalise_metadata(packet.meta)

        if not _same_epoch(self.epoch, meta["epoch"]):
            self.reset({"epoch": meta["epoch"], "reason": "new-epoch"})
        if (
            self.buffered_subframes > 0
            and meta["sequence"] != self._expected_sequence
        ):
            self._discard_partial()
        if (
            self.buffered_subframes > 0
            and meta["subframe_number"] != self.buffered_subframes
        ):
            self._discard_partial()

        if self.buffered_subframes == 0:
            if meta["subframe_number"] != 0:
                return Result.forward(message.clear_packet())
            self._allocate(packet.data)
            self._first_meta = meta

        assert self._buffer is not None
        column_start = int(meta["subframe_number"]) * 2
        for field in self._FIELDS:
            subframe = as_csi_4d(
                _mapping_field(packet.data, field, self._MATLAB_FIELDS[field]),
                name=f"data.{field}",
            )
            target = self._buffer[field]
            if subframe.shape[0] != target.shape[0] or subframe.shape[1] != 2:
                raise ValueError(f"data.{field} has incompatible subframe shape {subframe.shape}")
            if subframe.shape[2:] != target.shape[2:]:
                raise ValueError(f"data.{field} has incompatible antenna dimensions")
            target[:, column_start : column_start + 2, :, :] = subframe

        self.buffered_subframes += 1
        self._expected_sequence = meta["sequence"] + 1
        result = Result.forward(message.clear_packet())
        if meta["subframe_number"] != 9:
            return result

        assert self._first_meta is not None
        frame_meta = dict(self._first_meta)
        frame_meta["raw_end_sample_0"] = meta["raw_end_sample_0"]
        frame_meta["end_sequence"] = meta["sequence"]
        frame_meta.pop("subframe_number", None)
        output = Packet(
            type="csi-frame",
            data=self._buffer,
            meta=frame_meta,
            quality={"complete": True, "subframe_count": self.buffered_subframes},
        )
        result = Result.forward(self.replace_output(result.message, output))
        self.emitted_frames += 1
        self._clear_partial()
        return result

    def reset(self, event: Any = None) -> None:
        epoch = None
        if isinstance(event, dict):
            epoch = event.get("epoch", event.get("Epoch"))
        elif event is not None:
            epoch = event
        if self.buffered_subframes > 0:
            self.dropped_partial_frames += 1
        self.epoch = epoch
        self._clear_partial()

    def get_status(self) -> dict[str, Any]:
        return {
            "state": "ready",
            "ready": True,
            "epoch": self.epoch,
            "buffered_subframes": self.buffered_subframes,
            "emitted_frames": self.emitted_frames,
            "dropped_partial_frames": self.dropped_partial_frames,
        }

    def _allocate(self, data: Any) -> None:
        arrays = {
            field: as_csi_4d(
                _mapping_field(data, field, self._MATLAB_FIELDS[field]),
                name=f"data.{field}",
            )
            for field in self._FIELDS
        }
        for field, array in arrays.items():
            if array.shape[1] != 2:
                raise ValueError(f"data.{field} must contain two time samples")
        self._buffer = {
            field: complex_zeros_like(array, (array.shape[0], 20, array.shape[2], array.shape[3]))
            for field, array in arrays.items()
        }

    def _discard_partial(self) -> None:
        if self.buffered_subframes > 0:
            self.dropped_partial_frames += 1
        self._clear_partial()

    def _clear_partial(self) -> None:
        self._buffer = None
        self._first_meta = None
        self.buffered_subframes = 0
        self._expected_sequence = None

    @classmethod
    def _validate_packet(cls, packet: Packet) -> None:
        if not all(
            field in packet.data or cls._MATLAB_FIELDS[field] in packet.data
            for field in cls._FIELDS
        ):
            raise ValueError("Input is not a valid CSI subframe packet")
        _normalise_metadata(packet.meta)


def _mapping_field(mapping: Any, python_name: str, matlab_name: str) -> Any:
    if python_name in mapping:
        return mapping[python_name]
    if matlab_name in mapping:
        return mapping[matlab_name]
    raise ValueError(f"Missing data field {python_name!r}")


def _normalise_metadata(metadata: Any) -> dict[str, Any]:
    aliases = {
        "epoch": "Epoch",
        "sequence": "Sequence",
        "subframe_number": "SubframeNumber",
        "raw_end_sample_0": "RawEndSample0",
    }
    result: dict[str, Any] = {}
    for python_name, matlab_name in aliases.items():
        if python_name in metadata:
            result[python_name] = metadata[python_name]
        elif matlab_name in metadata:
            result[python_name] = metadata[matlab_name]
        else:
            raise ValueError("CSI subframe metadata is incomplete")
    result.update(metadata)
    return result


def _same_epoch(left: Any, right: Any) -> bool:
    if left == right:
        return True
    return (
        isinstance(left, (float, np.floating))
        and isinstance(right, (float, np.floating))
        and math.isnan(left)
        and math.isnan(right)
    )


__all__ = ["CsiFrameAssembler"]
